File: eel-core/src/main/scala/io/eels/component/json/JsonSink.scala
package io.eels.component.json

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
import io.eels.schema.StructType
import io.eels.{Rec, Sink, SinkWriter}
import org.apache.hadoop.fs.{FileSystem, Path}

case class JsonSink(path: Path)(implicit fs: FileSystem) extends Sink {

  override def open(schema: StructType): SinkWriter = new SinkWriter {

    private val lock = new AnyRef()
    private val out = fs.create(path)

    val mapper = new ObjectMapper with ScalaObjectMapper
    mapper.registerModule(DefaultScalaModule)

    override def write(row: Rec) {
      val map = schema.fieldNames.zip(row).toMap
      val json = mapper.writeValueAsString(map)
      lock.synchronized {
        out.writeBytes(json)
        out.writeBytes("\n")
      }
    }

    override def close() {
      out.close()
    }
  }
}

File: eel-core/src/main/kotlin/io/eels/component/json/JsonSink.kt
package io.eels.component.json

import io.eels.Row
import io.eels.schema.Schema
import io.eels.Sink
import io.eels.SinkWriter

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import org.codehaus.jackson.map.ObjectMapper

data class JsonSink(val path: Path) : Sink {

  override fun writer(schema: Schema): SinkWriter = object : SinkWriter {

    val lock = Any()
    val fs = FileSystem.get(Configuration())
    val out = fs.create(path)

    val mapper = ObjectMapper()

    override fun write(row: Row) {
      val map = schema.fieldNames().zip(row.values).toMap()
      val json = mapper.writeValueAsString(map)
      synchronized(lock) {
        out.writeBytes(json)
        out.writeBytes("\n")
      }
    }

    override fun close() {
      out.close()
    }
  }
}
